### PR TITLE
Validate GraphQL schema early

### LINKED ISSUE #579: Validate schema early
Currently graphql-core doesn't validate the schema until execution time (to allow for partial schema generation). Since strawberry doesn't support partial generation at the moment it would make sense to validate the schema as soon as it's instantiated so that any issues are raised early.

~~As a bonus it would be nice to be able to specify which validators to run.~~ Actually the schema validation is separate to the validate "rules" that I was thinking about here so this is not the right place to tackle that problem.

### PR DESCRIPTION
**Pull request by jkimbo**
## Description

Resolves #579 

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* #579 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/),so make sure to pick the appropriate type. If in doubt feel free to ask :)

**Review by patrick91 — COMMENTED** [review]
Uh! ✨  nice one! I'm not sure about the formatting of the errors, but we can deal with that later! I've tried rust this weekend and I loved their errors, so maybe we can get some inspiration from them, but again, let's deal with that in future :)

I have only one question, now that we validate the schema at creation time, shall we remove the validation from here: 
https://github.com/strawberry-graphql/strawberry/blob/e10e00070abb80bb17015c3481ea9ae47d269d41/strawberry/schema/execute.py#L45 ?

**Review by BryceBeagle — APPROVED** [review]

File: strawberry/schema/schema.py [hunk 1]
```diff
@@ -51,6 +51,13 @@ def __init__(
             types=[get_object_type(type, self.type_map) for type in types],
         )
 
+        # Validate schema early because we want developers to know about
+        # possible issues as soon as possible
+        errors = validate_schema(self._schema)
+        if errors:
+            formatted_errors = "\n\n".join(f"❌ {error.message}" for error in errors)
```
Comment (Ambro17) [inline strawberry/schema/schema.py:58]: +1 for the emoji

**Review by Ambro17 — APPROVED** [review]

**Comment by jkimbo** [comment]
> Uh! ✨ nice one! I'm not sure about the formatting of the errors, but we can deal with that later! I've tried rust this weekend and I loved their errors, so maybe we can get some inspiration from them, but again, let's deal with that in future :)

Yeah I would love to have some really rich error messages that can help developers find and resolve the issue quickly. From a visual perspective I've found the stack traces in poetry to be really nice. I think this is the code that formats the exception traces: https://github.com/sdispater/clikit/blob/master/src/clikit/ui/components/exception_trace.py

> I have only one question, now that we validate the schema at creation time, shall we remove the validation from here:
> https://github.com/strawberry-graphql/strawberry/blob/e10e00070abb80bb17015c3481ea9ae47d269d41/strawberry/schema/execute.py#L45

Didn't realise that the execute functions where inlined into Strawberry! I'll remove that line (though it doesn't actually matter because the validation only ever gets run once: https://github.com/graphql-python/graphql-core/blob/7d826f0ec0a447cb869dfb891a755dbe8ea9b66f/src/graphql/type/validate.py#L45-L73)

**Comment by jkimbo** [comment]
@patrick91 ready for review again

File: strawberry/schema/execute.py [hunk 2]
```diff
@@ -42,14 +41,8 @@ async def execute(
...
-        schema_validation_errors = validate_schema(schema)
-
-        if schema_validation_errors:
-            return ExecutionResult(
-                data=None,
-                errors=schema_validation_errors,
-                extensions=extensions_runner.get_extensions_results(),
-            )
+        # Note: In graphql-core the schema would be validated here but in
+        # Strawberry we are validating it at initialisation time instead
```
Comment (patrick91) [inline strawberry/schema/execute.py:45]: 💯

**Review by patrick91 — APPROVED** [review]
Awesome!

**Pull request merged**